src/Compilers/CSharp/Portable/SourceGeneration/IndentingStringBuilder.cs

```diff
@@ -0,0 +1,497 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.CodeDom.Compiler;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+/// <summary>
+/// A helper type for generating C# source code efficiently.  Content written is added to a <see cref="StringBuilder"/>,
+/// with default behavior provided for controlling the current indent level of the code.
```

Comment:
Can we make this write directly to `SourceText` so it both:

1. Doesn't have LOH allocations
2. Doesn't require conversion from string to SourceText for use by Roslyn

Reply:
Yes.  That's part of the original proposal.  I don't have that here though as this *currently* sits at a layer that doesn't have access to SourceText.   The final form would.